runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,225 @@
+  end
+  Watcher.stop_notifications()
+  Watcher.start_notifications()
+end
+
+-- Checks if a buffer should have a watcher attached to it.
+local function valid_buf(bufnr)
+  if bufnr < 0 then
+    return false
+  end
+
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  if not buflisted or buftype == 'nofile' or buftype == 'quickfix'
+    or not fname then
+    return false
+  end
+  return true
+end
+
+--- Creates and initializes a new watcher object with the given filename.
+---
+--@param bufnr: (required, number) The buffer number of the buffer that
+--- is to be watched.
+function Watcher:new(bufnr)
+  vim.validate{bufnr = {bufnr, 'number', false}}
+  -- get full path name for the file
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {bufnr = bufnr, fname = fname, ffname = ffname,
+       handle = nil, pending_notifs = false}
+  w._start_handle = function(_) end
+  setmetatable(w, self)
+  return w
+end
+
+--- Starts the watcher
+---
+--@param self: (required, table) The watcher which should be started.
+function Watcher:start()
+  set_mechanism('local', self.bufnr)
+  self._start_handle(self.bufnr)
+end
+
+--- Stops the watcher and closes the handle.
+---
+--@param self: (required, table) The watcher which should be stopped.
+function Watcher:stop()
+  if self.handle == nil then
+    return
+  end
+
+  self.handle:stop()
+
+  -- close the handle altogether, for windows.
+  if self.handle:is_closing() then
+    return
+  end
+  self.handle:close()
+end
+
+--- Debounces the watcher. Completely closes the handle and starts again with a new
+--- handle. Necessary for cases when the file being monitored is edited via editors
+--- like vim, which, depending on the value of `backupcopy`, move the original file
+--- for backing it up.
+---
+--@param self:(required, table) The watcher object which should be debounced.
+function Watcher:debounce()
+  self:stop()
+  self:start()
+end
+
+--- Callback for watcher handle. Marks a watcher as having pending
+--- notifications. The nature of notification is determined while
+--- responding to the notification.
+---
+--@param err: (string) Error if any occured during the execution of the callback.
+---
+function Watcher:on_change(err, _, _)
+  if err ~= nil then
+    error(err)
+  end
+
+  self.pending_notifs = true
+
+  self:debounce()
+end
+
+--- Starts and initializes a watcher for the given path. A thin wrapper around
+--- Watcher:start() that can be called from vimscript.
+---
+--@param bufnr: (required, string) The path that the watcher should watch.
+function Watcher.start_watch(bufnr)
+  bufnr = tonumber(bufnr)
+  if not valid_buf(bufnr) then
+    return
+  end
+
+  if WatcherList[bufnr] ~= nil then
+    WatcherList[bufnr]:stop()
+  end
+
+  WatcherList[bufnr] = Watcher:new(bufnr)
```

Comment:
Do we actually need to always create a new one? Can we just create a new one when there isn't one already?

Or will that not work?

Reply:
I didn't change this after I changed the indexing from filenames to buffer numbers. Since buffer numbers are unique, I don't think we should need to create a new watcher when one already exists. Thanks for pointing this out!